zwave_js_server/model/node/__init__.py

```diff
@@ -463,11 +465,26 @@ async def async_send_command(
         if require_schema is not None:
             kwargs["require_schema"] = require_schema
 
-        if wait_for_result or (
-            wait_for_result is None and self.status != NodeStatus.ASLEEP
-        ):
+        if wait_for_result:
             result = await self.client.async_send_command(message, **kwargs)
             return result
+        if wait_for_result is None and self.status not in (
+            NodeStatus.ASLEEP,
+            NodeStatus.DEAD,
+        ):
+            result_task = asyncio.create_task(
+                self.client.async_send_command(message, **kwargs)
+            )
+            status_task = asyncio.create_task(self._status_event.wait())
+            await asyncio.wait(
+                [result_task, status_task],
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+            status_task.cancel()
+            if self._status_event.is_set() and not result_task.done():
+                result_task.cancel()
+                return None
```

Comment:
Shouldn't we raise an error if the node died or fell asleep unexpectedly when waiting for a result?

Reply:
@AlCalzone what do you think? If at the point where we sent a command to the driver, the node status was awake or alive, but then while waiting for a response, the node status switched to asleep or dead, should we consider that unexpected behavior? Or should we expect that the driver will queue the command and will eventually issue it? I would think that if an error needed to be raised, it would be the Driver's responsibility to do so, no? The driver currently isn't doing that, so that may be a clue as to what the right answer is, but it's worth explicitly asking.

Likewise, if we must wait for a response from a command because the response is the primary purpose of issuing the command (e.g. checking for firmware update capabilities), what should happen when this scenario occurs? Should we continue waiting and hope the node comes back to an alive or awake state, or should we raise an error?

Reply:
The driver should not silently drop commands that expect a response. If it does, that's a driver bug - might have to do with the stalling firmware updates that have been reported recently.

> but then while waiting for a response, the node status switched to asleep or dead, should we consider that unexpected behavior?

IMO, no. A node can be assumed alive or awake, but fail to respond the moment we send a command. Especially with sleeping nodes, this has to be expected. If this is a mains-powered node, the driver should throw after giving up. If this is a node that can sleep, the driver should re-queue the command and eventually handle it.

Side note: I'm adding a callback applications can use to get notified about command progress updates:
https://github.com/zwave-js/node-zwave-js/pull/6212

Reply:
thanks so then I think the answer @MartinHjelmare is we don't want to raise, instead we should treat it like it's a sleeping/dead node and not wait for the result (except for in cases where we must wait for the result). If the node had been asleep or dead when the command was issued that would have been the behavior anyway right?

Reply:
Ok, good.